src/AgentImplementation.sol

```diff
@@ -55,40 +54,65 @@ contract AgentImplementation is IAgent, ERC721Holder, ERC1155Holder {
+    /// @param logics Array of logics to be executed
+    /// @param tokensReturn Array of ERC-20 tokens to be returned to the current user
+    function execute(IParam.Logic[] calldata logics, address[] calldata tokensReturn) external payable {
+        if (msg.sender != router) revert NotRouter();
+
+        _chargeFeeByMsgValue();
+
+        _executeLogics(logics, true);
+
+        _returnTokens(tokensReturn);
+    }
+
+    /// @notice Execute arbitrary logics and is only callable by the router using a signer's signature
+    /// @dev The router is designed to prevent reentrancy so additional prevention is not needed here
     /// @param logics Array of logics to be executed
     /// @param fees Array of fees
     /// @param tokensReturn Array of ERC-20 tokens to be returned to the current user
-    function execute(
+    function executeWithSignature(
         IParam.Logic[] calldata logics,
         IParam.Fee[] calldata fees,
         address[] calldata tokensReturn
-    ) external payable checkCaller {
-        _executeLogics(logics);
+    ) external payable {
+        if (msg.sender != router) revert NotRouter();
 
         _chargeFee(fees);
 
-        // Push tokensReturn if any balance
-        uint256 tokensReturnLength = tokensReturn.length;
-        if (tokensReturnLength > 0) {
-            address user = IRouter(router).currentUser();
-            for (uint256 i; i < tokensReturnLength; ) {
-                address token = tokensReturn[i];
-                if (token == _NATIVE) {
-                    payable(user).sendValue(address(this).balance);
-                } else {
-                    uint256 balance = IERC20(token).balanceOf(address(this));
-                    IERC20(token).safeTransfer(user, balance);
-                }
+        _executeLogics(logics, false);
 
-                unchecked {
-                    ++i;
-                }
-            }
+        _returnTokens(tokensReturn);
+    }
+
+    /// @notice Execute arbitrary logics and is only callable by a valid callback. Charge fee based on the scenarios
+    ///         defined in the router if the charge bit is set.
+    /// @dev A valid callback address is set during `_executeLogics` and reset here
+    /// @param logics Array of logics to be executed
+    function executeByCallback(IParam.Logic[] calldata logics) external payable {
+        bytes32 callbackWithCharge = _callbackWithCharge;
+
+        // Revert if msg.sender is not equal to the callback address
+        if (msg.sender != address(bytes20(callbackWithCharge))) revert NotCallback();
+
+        // Check the least significant bit to determine whether to charge fee
+        bool shouldChargeFeeByLogic = callbackWithCharge & _CHARGE_MASK != bytes32(0);
+
+        // Reset immediately to prevent reentrancy
+        _callbackWithCharge = _INIT_CALLBACK_WITH_CHARGE;
+
+        // Check if charge fee
+        if (shouldChargeFeeByLogic) {
```

Comment:
Can we eliminate the need for the `if .. else` statement and directly use `shouldChargeFeeByLogic` instead?

```suggestion
        _executeLogics(logics, shouldChargeFeeByLogic);
```

Reply:
Fixed in https://github.com/dinngo/composable-router-contract/pull/72/commits/a457a582df2e6abf44dc5a1cd79a0e669269d2f0. Thansk!